File: Sean的leetcode/python/Longest Increasing Path in a Matrix.py
###
# DFS
# Time Complexity: O(n*m) n:row m:col k:len of word
# Space Complexity: O(n*m)
###
class Solution(object):
    def longestIncreasingPath(self, matrix):
        """
        :type matrix: List[List[int]]
        :rtype: int
        """
        if not matrix or not matrix[0]:
            return 0
        rows = len(matrix)
        cols = len(matrix[0])
        max_len = 0
        dis = [[0] * cols for i in xrange(rows)]
        for i in xrange(rows):
            for j in xrange(cols):
                max_len = max(max_len, self.dfs(matrix, i, j, dis))
        return max_len
    
    def dfs(self, matrix, i, j, dis):
        rows = len(matrix)
        cols = len(matrix[0])
        if dis[i][j] != 0:
            return dis[i][j]
        dis[i][j] = 1
        direc = [(-1, 0), (1, 0), (0, 1), (0, -1)]
        for di, dj in direc:
            nbi = i + di
            nbj = j + dj
            if nbi >= 0 and nbi < rows and nbj >= 0 and nbj < cols and matrix[nbi][nbj] > matrix[i][j]:
                dis[i][j] = max(dis[i][j], self.dfs(matrix, nbi, nbj, dis) + 1)
        return dis[i][j]
        
        

File: MyLeetCode/Leetcode/Longest Increasing Path in a Matrix.cpp
/*
 DFS
 */
class Solution {
public:
    vector<vector<int>> dis;
    int dfs(vector<vector<int>>& matrix, int r, int c) {
        vector<pair<int, int>> dir = {{-1, 0}, {1, 0}, {0, -1}, {0, 1}};
        int rows = matrix.size();
        int cols = rows? matrix[0].size() : 0;
        
        if (dis[r][c] != 0) return dis[r][c];
        
        int curr = 1;
        for (auto d : dir) {
            int nb_r = r + d.first;
            int nb_c = c + d.second;
            if (nb_r >= 0 && nb_r < rows && nb_c >= 0 && nb_c < cols && matrix[nb_r][nb_c] > matrix[r][c]) {
                curr = max(curr, 1 + dfs(matrix, nb_r, nb_c));
            }
        }
        dis[r][c] = curr;
        
        return curr;
    }
    int longestIncreasingPath(vector<vector<int>>& matrix) {
        int rows = matrix.size();
        int cols = rows? matrix[0].size() : 0;
        if(!rows || !cols) return 0;
        dis.resize(rows, vector<int>(cols));
        int maxlen = 1;
        for (int i = 0; i < rows; i++) {
            for (int j = 0; j < cols; j++) {
                maxlen = max(maxlen, dfs(matrix, i, j));
            }
        }
        return maxlen;
        
    }
};